src/project-root.ts

```diff
@@ -1,73 +1,164 @@
-import * as vscode from 'vscode';
-import * as util from './utilities';
 import * as config from './config';
+import * as vscode from 'vscode';
 import * as path from 'path';
 
-export async function findProjectRoots(): Promise<vscode.Uri[]> {
-  const projectFileNames: string[] = [
-    'project.clj',
-    'shadow-cljs.edn',
-    'deps.edn',
-    'bb.edn',
-    '.nrepl-port',
-    'settings.gradle',
-    'settings.gradle.kts',
-  ];
+export type ProjectRoot = {
+  uri: vscode.Uri;
+  reason: string;
+};
+
+export const rootToUri = (root_or_uri: ProjectRoot | vscode.Uri) => {
+  if (root_or_uri instanceof vscode.Uri) {
+    return root_or_uri;
+  }
+  return root_or_uri.uri;
+};
+
+type FindRootParams = {
+  /**
+   * If set to `false` then root workspace folders won't be included in the list by default. They will only
+   * be included if they match the same project heuristics as any other folder in the workspace.
+   *
+   * @default true
+   */
+  include_workspace_folders?: boolean;
+};
+
+/**
+ * Returns a list of discovered project roots. A directory is considered a project root if it contains certain
+ * clojure related files that identify it as a Clojure project. For example, a `deps.edn` file.
+ *
+ * The response contains a `reason` as to why a folder is considered a project root which can be used in
+ * selection menus to help the user.
+ */
+export async function findProjectRootsWithReasons(params?: FindRootParams) {
+  const projectFileNames: string[] = ['project.clj', 'shadow-cljs.edn', 'deps.edn', 'bb.edn'];
```

Comment:
Is there a reason why this can't just be `findProjectRootsWithReasons(includeWorkspaceFolders = true)`?

That would be less to scan and wonder about. But if there's some reason why the type is needed, maybe the extra parsing is worth it.

Reply:
Generally I prefer to use maps with well-named keys in function parameters. It's slightly more verbose at the function definition, but makes it significantly easier to understand what is happening at the call-site of the function - especially when reviewing code.

Passing a boolean doesn't provide enough context to a reader without them having to look at the function implementation.

It also provides better auto-complete assistance to code authors who are using the function as the doc-strings are provided per-parameter instead of on the function overall.

Reply:
Makes sense. We should consider doing more of this in the codebase, probably.